### PR TITLE
improve update systems via recurring states

### PR DESCRIPTION
**Pull request by mcalmer**
## What does this PR change?

Implement some improvements to update systems via recurring states.

- provide a standalone salt update state
- improve update stack update state
- check pillar to skip reboot_if_needed if not wanted

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- Not needed, but added https://github.com/SUSE/spacewalk/issues/23573

- [x] **DONE**

## Test coverage
- No tests: already covered

- [x] **DONE**

## Links

Issue(s): https://github.com/SUSE/spacewalk/issues/22483
Ports(s): https://github.com/SUSE/spacewalk/pull/23577

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

### CODE REVIEW

**Review by rjmateus — APPROVED** [review]
Lgtm. thank you for implementing this one.

**Review by cbbayburt — COMMENTED** [review]
How is the `mgr_reboot_if_needed` pillar set?

**Comment by mcalmer** [comment]
> How is the `mgr_reboot_if_needed` pillar set?

I think there is no automatic way yet. So by default nothing change compared to what we have now. reboot is always executed if needed.

But a customer can manual set it via `/srv/pillar/` or maybe also via "custom info" (AFAIK we convert them into pillar? - I havn't tried this yet.)

**Review by cbbayburt — APPROVED** [review]
Thanks. LGTM :+1: 

Related issue just for reference: https://github.com/SUSE/spacewalk/issues/22483

**Comment by mcalmer** [comment]
When using custom info, it would look like this:
```
    custom_info:
        ----------
        mgr_reboot_if_needed:
            False
```
And I think "False" is here a string. I think we can enhance the check to also react on this.

**Comment by mcalmer** [comment]
I am also looking for some hints if the UI configutation is complete? AFAIK there is somewhere a decription which I cannot identify where it comes from.

**Review by cbbayburt — CHANGES_REQUESTED** [review]
> I am also looking for some hints if the UI configuration is complete? AFAIK there is somewhere a description which I cannot identify where it comes from.

Oh right, descriptions are translated strings in [`java/StringResource_en_US.xml`](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/frontend/strings/java/StringResource_en_US.xml#L9309-L9311).

The UI picks it up automatically, looking at the message key `internal_state.<state_label>`

So you need to add the following:

```xml
      <trans-unit id="internal_state.update-salt">
        <source>...</source>
      </trans-unit>
```

File: susemanager-utils/susemanager-sls/salt/uptodate.sls [hunk 1]
```diff
@@ -31,6 +40,7 @@ mgr_keep_system_up2date_pkgs:
       - sls: channels
       - pkg: mgr_keep_system_up2date_updatestack
 
+{%- if salt['pillar.get']('mgr_reboot_if_needed', True) %}
```
Comment (parlt91) [inline susemanager-utils/susemanager-sls/salt/uptodate.sls:43]: Just for understanding. The default value of an unset pillar is `True`? Couldn't find anything on it in the docs

Comment (mcalmer) [reply susemanager-utils/susemanager-sls/salt/uptodate.sls:43]: Yes, the `True` is the default value if this pillar does not exist
https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.pillar.html#salt.modules.pillar.get

**Comment by mcalmer** [comment]
@parlt91 @cbbayburt I added the translation and I have added handling of pillar created via custom_info.
Please check again

**Comment by mcalmer** [comment]
@admd should this be ported to 4.3 as well? If @parlt91 port the one shot execution for recurring states, this may make sense as well.

File: susemanager-utils/susemanager-sls/salt/uptodate.sls (hunk 1 above)
Comment (parlt91) [reply susemanager-utils/susemanager-sls/salt/uptodate.sls]: Now I understand. The `True` is the default not the expected value. Thank you

**Review by parlt91 — APPROVED** [review]
Tested it out and it looks good to me. Just left one comment about the schema upgrade dir above.

File: schema/spacewalk/upgrade/susemanager-schema-5.0.3-to-susemanager-schema-5.0.4/100-add-internal-state-update-salt.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+INSERT INTO suseInternalState (id, name, label)
```
[the remark this answers is not in the document]
Comment (mcalmer) [reply schema/spacewalk/upgrade/susemanager-schema-5.0.3-to-susemanager-schema-5.0.4/100-add-internal-state-update-salt.sql:1]: yesterday this dir was correct - today maybe not :-)

**Comment by cbbayburt** [comment]
> @admd should this be ported to 4.3 as well? If @parlt91 port the one shot execution for recurring states, this may make sense as well.

IMO yes. https://github.com/SUSE/spacewalk/issues/22483 is also going to be implemented on top of this.

**Review by cbbayburt — APPROVED** [review]

File: schema/spacewalk/upgrade/susemanager-schema-5.0.3-to-susemanager-schema-5.0.4/100-add-internal-state-update-salt.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcalmer) [reply schema/spacewalk/upgrade/susemanager-schema-5.0.3-to-susemanager-schema-5.0.4/100-add-internal-state-update-salt.sql]: moved

**Comment by admd** [comment]
Generally, we plan not to add anything major in 4.3 anymore.

But in this case, this would be a nice change to add in 4.3. It's not that big from code wise so less chance of introducing any regressions.

**Comment by rjmateus** [comment]
@admd @mcalmer @cbbayburt I was discussion this with MC on the chat. COnsidering that update salt with salt is the a root of problems, and MC (thank you so much) added this state to address the issue, would make sense to keep updating salt on the `mgr_keep_system_up2date_updatestack` state? [1]
If the user uses this new state before update, we would not have have any problems. However, if salt gets updated, it restarts the agent and forgets about the remaining tasks.

https://github.com/uyuni-project/uyuni/blob/master/susemanager-utils/susemanager-sls/salt/uptodate.sls#L1-L26

**Pull request merged**